### PR TITLE
Float interface changes

### PR DESCRIPTION
**Pull request by aijony**
- Default to silent errors for floats (e.g. propagate NaNs and infinities)
  - Access to `nan` and `infinity` with `Transfinite`
- Rename `Float` to [`Elementary`](https://en.wikipedia.org/wiki/Elementary_function)
  - Rename `expt` to `pow` and define it in terms of `ln` and `exp`
- Add `Radical` for [algebraic](https://en.wikipedia.org/wiki/Algebraic_number) operations
  - Add `nth-root`
- Move to `math/elementary`
- Imply (principal) complex elementary functions
  - Add `Polar` class
 
Requires:  #594

### CODE REVIEW

File: library/math/elementary.lisp [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+    - (log b x) = (/ (ln x) (ln b))
+    - (pow x y) = (exp (* y (ln x)))
+"
+    (exp (:a -> :a))
+    (pow (:a -> :a -> :a))
+    (ln  (:a -> :a))
+    (log  (:a -> :a -> :a)))
+
+  (define-class (Radical :a)
+    "Obeys, (^ (sqrt x) 2) = x and (^^ (nth-root n x) n) = x"
```
Comment (stylewarning) [inline library/math/elementary.lisp:69]: no comma after "obeys"

File: library/math/elementary.lisp [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+    (nth-root (Integer -> :a -> :a))
+    (sqrt (:a -> :a)))
+
+  (define-class ((Complex :a) (Num :a) => Polar :a)
+    "For a complex number z = (complex x y)
+    z = (* (magnitude z) (exp (* ii (phase z))))
```
Comment (stylewarning) [inline library/math/elementary.lisp:75]: you might want to format these as markdown

File: library/math/elementary.lisp [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+      (cond
+        ;; y /= 0
+        ((or (/= 0 (floor y)) (/= 0 (ceiling y)))
+         (* 2 (atan (/ (- r x) y))))
+        ;; x > 0
+        ((> (floor x) 0) 0)
+        ;; x < 0
+        ((< (ceiling x) 0) pi)
+        ;; x = 0, y = 0
+        ;; Pretend principle angle of 0 is 0
```
Comment (stylewarning) [inline library/math/elementary.lisp:156]: "principal"

File: library/math/elementary.lisp [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+      ((Tuple (EQ) (GT)) (/ pi 2))
+      ((Tuple (EQ) (LT)) (/ pi -2))
+      ((Tuple (EQ) (EQ)) 0)))
+
+  (define (canonical-nth-root n x)
+    "By definition implementation of `nth-root' for reals"
+    (pow x (reciprocal (fromInt n))))
+
+  (define (canonical-polar z)
+    "By definition implementation of `polar'"
```
Comment (stylewarning) [inline library/math/elementary.lisp:142]: > The by-definition implementation of `polar`.

File: library/math/elementary.lisp [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+    (match (Tuple (<=> x 0) (<=> y 0))
+      ((Tuple (GT) _)    (atan (/ y x)))
+      ((Tuple (LT) (LT)) (- (atan (/ y x)) pi))
+      ((Tuple (LT) _)    (+ (atan (/ y x)) pi))
+      ((Tuple (EQ) (GT)) (/ pi 2))
+      ((Tuple (EQ) (LT)) (/ pi -2))
+      ((Tuple (EQ) (EQ)) 0)))
+
+  (define (canonical-nth-root n x)
+    "By definition implementation of `nth-root' for reals"
```
Comment (stylewarning) [inline library/math/elementary.lisp:138]: > The by-definition implementation of `nth-root` for real numbers.

File: library/math/elementary.lisp [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+         (match a
+           ((%Complex a _) a)))
+       (define (imag-part a)
+         (match a
+           ((%Complex _ b) b))))))
+
+(coalton-toplevel
+  (define-instance ((Elementary :a) => (Exponentiable (Complex :a)))
+    (define (ln z)
+      ;; The principle natural log of a complex number
```
Comment (stylewarning) [inline library/math/elementary.lisp:282]: principal

File: library/math/arith.lisp [hunk 7]
```diff
@@ -446,199 +398,115 @@ The function general/ is partial, and will error produce a run-time error if the
...
+      (lisp Double-Float ()
+        float-features:double-float-nan))
+    (define (nan? x)
+      (Lisp Boolean (x)
+        (float-features:float-NaN-p x)))
+    (define (infinite? x)
+      (Lisp Boolean (x)
+        (float-features:float-infinity-p x))))
+
+  (define-instance (Into Integer (Optional Single-Float))
```
Comment (stylewarning) [inline library/math/arith.lisp:441]: do we want `TryInto` instances (too)?

**Review by stylewarning — APPROVED** [review]
some typos, but looks good overall...

we should probably beef up testing more for all of this numerical code, but i'm fine for that to not delay this PR

File: library/math/arith.lisp (hunk 7 above)
Comment (eliaslfox) [reply library/math/arith.lisp:441]: These should be `TryInto` instead of `Into`.

File: library/math/elementary.lisp (hunk 2 above)
Comment (aijony) [reply library/math/elementary.lisp:75]: I agree, but I am going to wait until after #612 so I can see the output.

**Pull request merged**